### PR TITLE
Bugfix signig service doesnt detect when connection is dropped (no heartbeat)

### PR DESCRIPTION
**Pull request by dybi**
resolves: https://github.com/golemfactory/concent/issues/861

### CODE REVIEW

File: signing_service/signing_service/constants.py [hunk 1]
```diff
@@ -38,3 +38,9 @@
 
 # Defines how much time in seconds should SigningService wait for AuthenticationChallengeFrame.
 RECEIVE_AUTHENTICATION_CHALLENGE_TIMEOUT = 20
+
+# Defines how many times Signing Service should try reconnecting on socket error, before giving up and crashing.
+SIGNING_SERVICE_DEFAULT_RECONNECT_ATTEMPTS = 5
```
Comment (rwrzesien) [inline signing_service/signing_service/constants.py]: Why should it crash instead of constantly retrying?

File: signing_service/signing_service/signing_service.py [hunk 2]
```diff
@@ -96,9 +102,18 @@ def __init__(
         self.ethereum_private_key = ethereum_private_key  # type: str
         self.current_reconnect_delay: Union[int, None] = None
         self.was_sigterm_caught: bool = False
+        self.reconnection_counter = 0
+        self.maximum_reconnection_attempts = maximum_reconnect_attempts
 
         self._validate_arguments()
 
+        def _set_was_sigterm_caught_true(signum: int, frame: Optional[FrameType]) -> None:  # pylint: disable=unused-argument
```
Comment (rwrzesien) [inline signing_service/signing_service/signing_service.py]: Why was this moved here?

Comment (dybi) [reply signing_service/signing_service/signing_service.py]: It was a part of a small refactor of `Signing Service` - -> you need to set it only once, and `run()` can be called multiple times, contrary to '__init__', so I figured out, that would be the better place for signal registration.

File: signing_service/signing_service/constants.py (hunk 1 above)
Comment (dybi) [reply signing_service/signing_service/constants.py]: I have had an interesting chat with @cameel and we finally got to a conclusion, that there should be retry limit (and when it is exceeded, application should crash & report to `Sentry` to get some attention)

File: signing_service/signing_service/signing_service.py (hunk 2 above)
Comment (rwrzesien) [reply signing_service/signing_service/signing_service.py]: It actually doesn't do anything before `run()` is called but ok. Please just fix indentation.

File: signing_service/signing_service/constants.py (hunk 1 above)
Comment (Jakub89) [reply signing_service/signing_service/constants.py]: > Why should it crash instead of constantly retrying?

We are not gone know when MM crashed if it just retrying. But 5 steps is to short of a time I think.

**Comment by cameel** [comment]
MIlestone should be unset until the pull request is merged.

Please read https://github.com/Code-Poets/coding-style/issues/19

File: signing_service/signing_service/signing_service.py [hunk 3]
```diff
@@ -96,9 +101,18 @@ def __init__(
...
+            logger.info('Closing connection and exiting on SIGTERM.')
+            self.was_sigterm_caught = True
```
Comment (Jakub89) [inline signing_service/signing_service/signing_service.py]: ************* Module signing_service.signing_service
W:110, 0: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
W:111, 0: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)

File: signing_service/signing_service/tests/test_run.py [hunk 4]
```diff
@@ -76,23 +78,19 @@ def test_that_signing_service_should_reraise_unrecognized_exception(self):
...
-                signing_service = SigningService(*self.parameters)
-                with pytest.raises(socket.error):
-                    signing_service.run()
-
-        mock_socket_connect.assert_called_once_with(('127.0.0.1', self.port))
-        mock_socket_close.assert_called_once()
+                with mock.patch('signing_service.signing_service.sleep'):
+                    signing_service = SigningService(*self.parameters)
+                    with pytest.raises(SigningServiceMaximumReconnectionAttemptsExceeded):
+                        signing_service.run()
```
Comment (Jakub89) [inline signing_service/signing_service/tests/test_run.py]: ************* Module signing_service.tests.test_run
W: 94, 0: Bad indentation. Found 28 spaces, expected 24 (bad-indentation)

File: middleman_protocol/middleman_protocol/tests/test_stream.py [hunk 5]
```diff
@@ -392,3 +392,20 @@ def test_that_raising_error_in_generator_should_call_close_on_socket(self, unuse
                             next(split_stream_generator)
 
                 mock_socket_close.assert_called_once()
+
+    def test_that_when_socket_receives_no_bytes_socket_error_is_raised(self, unused_tcp_port):
```
Comment (Jakub89) [inline middleman_protocol/middleman_protocol/tests/test_stream.py]: ************* Module middleman_protocol.tests.test_stream
R:396, 4: Method could be a function (no-self-use)
